### ISSUE TITLE
Provide a way to extract intermediate gradients

### ISSUE THREAD

**Issue by dionhaefner** [pasteurlabs/tesseract-jax#133]
### Summary

When chaining multiple ops and calling `jax.grad` or similar we typically lose the ability to inspect / debug the results (primals / Jacobians / JVPs / VJPs) of each step. This hinders interactive debugging while developing pipelines.

### Why is this needed?

Debugging end-to-end gradient pipelines is hard.

### Usage example

Here introducing a new function `sow` (inspired by [`flax.sow`](https://flax-linen.readthedocs.io/en/latest/guides/model_inspection/extracting_intermediates.html#store-intermediate-values-in-a-new-variable-collection)) to store intermediate values.

```python
from tesseract_jax import apply_tesseract, sow, run_pipeline

@jax.jit
def my_pipeline(inputs):
    res = apply_tesseract("mytess", inputs)
    sow(res, "mytess_result)
    res = apply_tesseract("yourtess", inputs)
    sow(res, "yourtess_result")
    return res

pipeline_grad = jax.grad(my_pipeline)
res, intermediates = run_pipeline(pipeline_grad, capture_intermediates=True)
```

This isn't the only plausible design for this; we could also rely on JAX built-ins from `jax.debug` or log stuff as a side-effect of `apply_tesseract` (maybe easiest to implement).
